rate-limiter/src/rate_limiter.rs

```diff
@@ -20,25 +23,36 @@ impl Clone for SleepingRateLimiter {
     }
 }
 
+impl From<Arc<Mutex<TokenBucket>>> for SleepingRateLimiter {
+    fn from(rate_limiter: Arc<Mutex<TokenBucket>>) -> Self {
+        Self { rate_limiter }
+    }
+}
+
 impl SleepingRateLimiter {
     /// Constructs a instance of [SleepingRateLimiter] with given target rate-per-second.
     pub fn new(rate_per_second: usize) -> Self {
         Self {
-            rate_limiter: TokenBucket::new(rate_per_second),
+            rate_limiter: Arc::new(Mutex::new(TokenBucket::new(rate_per_second))),
         }
     }
 
     /// Given `read_size`, that is an amount of units of some governed resource, delays return of `Self` to satisfy configure
     /// rate.
-    pub async fn rate_limit(mut self, read_size: usize) -> Self {
+    pub async fn rate_limit(self, read_size: usize) -> Self {
         trace!(
             target: LOG_TARGET,
             "Rate-Limiter attempting to read {}.",
             read_size
         );
 
-        let now = Instant::now();
-        let delay = self.rate_limiter.rate_limit(read_size, now);
+        let delay = {
+            let mut rate_limiter_lock = match self.rate_limiter.lock() {
```

Comment:
Won't this become a bottleneck? You're now taking a lock for every "packet"?

Reply:
I did some benchmarks. It wasn't a bottleneck, but actually I used wrong type of mutex - we prefer version from `parking-lot`.